File: csharp/Tachyon.Sdk/CollectionsResource.cs
namespace Tachyon.Sdk;

/// <summary><c>/collections</c> — create, list, and remove collections.</summary>
public sealed class CollectionsResource
{
    private readonly HttpTransport _transport;

    internal CollectionsResource(HttpTransport transport) => _transport = transport;

    /// <summary><c>POST /collections</c>. Field types are immutable after creation.</summary>
    public Task<CollectionInfo> CreateAsync(CollectionSchema schema, CancellationToken cancellationToken = default) =>
        _transport.RequestAsync<CollectionInfo>(HttpMethod.Post, "/collections", body: schema, cancellationToken: cancellationToken);

    /// <summary><c>GET /collections</c>.</summary>
    public Task<IReadOnlyList<CollectionInfo>> ListAsync(CancellationToken cancellationToken = default) =>
        _transport.RequestAsync<IReadOnlyList<CollectionInfo>>(HttpMethod.Get, "/collections", cancellationToken: cancellationToken);

    /// <summary><c>GET /collections/{name}</c>.</summary>
    public Task<CollectionInfo> RetrieveAsync(string name, CancellationToken cancellationToken = default) =>
        _transport.RequestAsync<CollectionInfo>(HttpMethod.Get, $"/collections/{Uri.EscapeDataString(name)}", cancellationToken: cancellationToken);

    /// <summary><c>DELETE /collections/{name}</c>. Removes the collection and all its data.</summary>
    public Task DeleteAsync(string name, CancellationToken cancellationToken = default) =>
        _transport.RequestAsync(HttpMethod.Delete, $"/collections/{Uri.EscapeDataString(name)}", cancellationToken: cancellationToken);
}

File: java/src/main/java/io/github/adikeshri/tachyon/CollectionsResource.java
package io.github.adikeshri.tachyon;

import com.fasterxml.jackson.core.type.TypeReference;

import java.util.List;

/** {@code /collections} — create, list, and remove collections. */
public final class CollectionsResource {
    private final HttpTransport transport;

    CollectionsResource(HttpTransport transport) {
        this.transport = transport;
    }

    /** {@code POST /collections}. Field types are immutable after creation. */
    public CollectionInfo create(CollectionSchema schema) {
        return transport.request("POST", "/collections", null, schema, CollectionInfo.class);
    }

    /** {@code GET /collections}. */
    public List<CollectionInfo> list() {
        return transport.request("GET", "/collections", null, null, new TypeReference<List<CollectionInfo>>() {
        });
    }

    /** {@code GET /collections/{name}}. */
    public CollectionInfo retrieve(String name) {
        return transport.request("GET", "/collections/" + UrlEncoding.pathSegment(name), null, null, CollectionInfo.class);
    }

    /** {@code DELETE /collections/{name}}. Removes the collection and all its data. */
    public void delete(String name) {
        transport.requestVoid("DELETE", "/collections/" + UrlEncoding.pathSegment(name));
    }
}
